### PR TITLE
Add accessibilityElementsHidden Prop To View Comp.

### LINKED ISSUE #9725: Accessibility with NavigationCardStack speaks accessibilityLabels of previous pages's views on iOS
### Issue Description

Let's say we have two pages in our app: Page A and Page B. NavigatorExperimental is used to navigate between them.

Page A's accessibility is working great. But when moving two Page B, VoiceOver still recognises and speaks Page A's accessible elements, even though they are now hidden.
### Steps to Reproduce / Code Snippets

Create 2 pages: Page A and Page B, and use NavigationExperimental to navigate between them. Both pages should contain some accessible elements.

Turn on VoiceOver on the iOS device and start the app. Navigate to Page B and try VoiceOver.
#### Expected Results

Page A's accessible elements should not appear in Page B.
### Additional Information
- React Native version: 0.32.0
- Platform(s) (iOS, Android, or both?): iOS
- Operating System (macOS, Linux, or Windows?): macOS

### PR DESCRIPTION
**Pull request by omeid**
accessibilityElementsHidden is used to control wheter elemetns contained
within a view are hidden from accessiblity services that query the
screen.

On iOS, it simply reflects UIKit's accessibilityElementsHidden, on
Android it is implemented with importantForAccessibility.

This commit also improves accessiblity of Navigation Experimental
CardStack by marking inactive (off screen) scenes invsible.

Closes #9725

### CODE REVIEW

**Comment by RubenSandwich** [comment]
Previous discussion on this issue happened on this [pull request](https://github.com/facebook/react-native/pull/10918). It looks like @omeid created a new PR because the previous pull request was languishing. This PR is important to me because without it the Navigator Experimental is unaccessible on iOS.

**Comment by omeid** [comment]
@RubenSandwich Yes, that pull request had to be rebased so many times I somehow missed some files in the process and it was rather messy, so I made this one.

**Comment by lacker** [comment]
I think that breaking the current `importantForAccessibility` behavior is too much of a cost to pay for this. Would it be possible to just make `importantForAccessibility` work on iOS as well? Another issue is that we intend to deprecate NavigationExperimental, so we are trying to avoid adding any more functionality there.

**Comment by omeid** [comment]
Hi @lacker,

This is not really breaking `importantForAccessiblity`, it marks it deprecated in favour of `accessiblityElementsHidden`.

However, as per initial conversation about this feature, if there is reasons to keep `importantForAccessiblity`, we should consider how we go about it.

re NavigationExperimental, this is not really doing much there, it is just making sure the compoent is not _totally_ inaccessable, which is a big shame.

**Comment by lacker** [comment]
Well, marking it as deprecated is implying that we'll remove it eventually, so it's introducing a breaking change, even if the day it breaks is not today.

**Comment by omeid** [comment]
My understanding is that deprecation is implying the intention of removing something; and this phase exists to ease the transition and provide an oppurtunity for reconsideration.

Regardless of semantics, as someone who is using this feature in both Android and iOS I would be interested to know what is the usecase for `importantForAccessiblity` that isn't possible with `accessiblityElementsHidden` and see how we can accommodate said requirments in a platform consistent way. Please also see #10918 for a bit of background.

Also, thanks for looking into this issue. It is something that I would really love to get fixed upstream so I don't have to manage my local fork. :(

**Comment by lacker** [comment]
The main use case for `importantForAccessibility` I am thinking about here is just not breaking peoples' current applications. We started making the accessibility API in this way, so unless there's a really good reason we should keep doing it in this way. Would it be possible to just make it work on iOS without doing a deprecation / migration sort of thing?

**Comment by omeid** [comment]
The reason for that is because `importantForAccessiblity` is a superset of `accessiblityElementsHidden` so impossible to make it crossplatfrom in a consistent way.

In detail; It is not possible to have a consistent `importantForAccessiblity` because on iOS, `accessiblityElementsHidden` is a boolean which corsponds to Android's `importantForAccessiblity` values `yes` (accessibilityElementsHidden=false) and `noHideDescendants` (accessibilityElementsHidden=true). However, iOS has no equivalent for `importantForAccessiblity=auto` and `importantForAccessiblity=no`.

**Comment by lacker** [comment]
Hmm. OK, unfortunately that leaves me in the situation of being strongly suspicious this is not the right way to solve the problem, but not being sure what the right way is to give you some helpful direction.

**Comment by omeid** [comment]
It may not be the prefect solution but I doubt anyone is using `auto` or `no`, or that their requirement couldn't be handled with `yes` and `noHideDescendants` which is provided by `accessiblityElementsHidden`.

Regardless, It is worth considering that as it stands, every react-naitve app that use navigation-experimental on iOS is embrassingly unaccessable from a VoiceOver point of view, to draw a parallel, it is so bad that is on par with having  `opacity: 0.5` set for every component from a visual point of view.

**Comment by sdg9** [comment]
@lacker regarding your [previous comment](https://github.com/facebook/react-native/pull/11788#issuecomment-278509000)

> Another issue is that we intend to deprecate NavigationExperimental, so we are trying to avoid adding any more functionality there.

That makes sense but I believe the problem is bigger than NavigationExperimental.  Because of your above comments I tried [react-navigation](https://github.com/react-community/react-navigation) and found it too has [the same accessibility problem](https://github.com/react-community/react-navigation/issues/377).  

Since iOS doesn't have an equivalent to android's importantForAccessibility prop exposed to the js layer I don't think these navigation libraries can do anything about this short of rendering one screen at a time.

My understanding is that while react-navigation is still in beta it's going to be the recommended approach going forward for navigation.  If so this PR or something similar will be required to make ADA compliant apps on iOS.  

If not this approach do you have any suggestions on which approach to take?  Expose `accessibilityElementsHidden` for iOS only and leave `importantForAccessibility` android only?

**Comment by satya164** [comment]
> The main use case for importantForAccessibility I am thinking about here is just not breaking peoples' current applications.

Generally I agree, but seems deprecating looks like the best option here. We usually have almost 3-4 months before finally removing the feature, so I believe people will have enough time to do it. Also isn't this just a quick search and replace?

> If not this approach do you have any suggestions on which approach to take? Expose accessibilityElementsHidden for iOS only and leave importantForAccessibility android only?

I think it's not good to introduce more fragmentation in the APIs.

**Comment by rigdern** [comment]
@omeid Are you sure that your implementation of `accessibilityElementsHidden` behaves the same on iOS and Android? I see a couple of potential inconsistencies:
  - The doc comment for the `accessibilityElementsHidden` prop says `true` means that the view and all of its descendants are hidden. However, [Apple's docs](https://developer.apple.com/reference/objectivec/nsobject/1615080-accessibilityelementshidden) say `true` hides the descendants but not the view.
  - The doc comment for the `accessibilityElementsHidden` prop makes it sound like `null` (default) means the OS uses heuristics to determine whether or not the view should be hidden. However, [Apple's docs](https://developer.apple.com/reference/objectivec/nsobject/1615080-accessibilityelementshidden) say the default value is `false`.

Do you know how iOS determines what is visible to the screen reader? For example, is a View with opacity 0 visible to the screen reader? As of commit https://github.com/facebook/react-native/commit/54beee261667539d0314abc6bc39363784e41e37 (RN 39) the NavigationExperimental now sets the opacity of the previous page to 0 (previously it was 0.3). The commit was intended to reduce overdraw on Android but I'm wondering if it also happened to make the previous page invisible to the iOS and Android screen readers.

**Comment by omeid** [comment]
@rigdern 

> The doc comment for the accessibilityElementsHidden prop says true means that the view and all of its descendants are hidden. However, Apple's docs say true hides the descendants but not the view.

The documentation could be improved, however it is not inconsistent from Android, as we use `importantForAccessiblity=noHideDescendants` on Android, which is functionally identical.

> The doc comment for the accessibilityElementsHidden prop makes it sound like null (default) means the OS uses heuristics to determine whether or not the view should be hidden. However, Apple's docs say the default value is false.

Android does some heuristics, for example --it considers opacity-- but iOS does not, it is still the OS decisions though.

> Do you know how iOS determines what is visible to the screen reader? For example, is a View with opacity 0 visible to the screen reader? As of commit 54beee2 (RN 39) the NavigationExperimental now sets the opacity of the previous page to 0 (previously it was 0.3). The commit was intended to reduce overdraw on Android but I'm wondering if it also happened to make the previous page invisible to the iOS and Android screen readers.


On Android, opacity 0 actually does help with accessibility as the default `importantForAccessiblity=auto` considers it, on iOS, it makes no difference.

**Comment by rigdern** [comment]
@omeid, thanks for the response.

> > The doc comment for the accessibilityElementsHidden prop says true means that the view and all of its descendants are hidden. However, Apple's docs say true hides the descendants but not the view.

> The documentation could be improved, however it is not inconsistent from Android, as we use importantForAccessiblity=noHideDescendants on Android, which is functionally identical.

After looking at the iOS and Android docs, it sounds to me like the prop works differently on iOS and Android for `accessibilityElementsHidden=true`:
  - iOS: The descendants are hidden
  - Android: The descendants and the view are hidden

Do you agree?

**Comment by omeid** [comment]
> After looking at the iOS and Android docs, it sounds to me like the prop works differently on iOS and > Android for accessibilityElementsHidden=true:
>
>  -   iOS: The descendants are hidden
>  -   Android: The descendants and the view are hidden
>
> Do you agree?

No. Please see the implementation; on [Android](https://github.com/facebook/react-native/pull/11788/files#diff-a73c2a0c22eceb34cf86aaede40f931fR106) `accessibilityElementsHidden=true` corresponds to `importantForAccessiblity=noHideDescendants`, on iOS it corresponds to `accessibilityElementsHidden=true`, the fact that it has the same name on iOS is merely conventional.

**Comment by rigdern** [comment]
@omeid I'm confused about which part of my statement you disagree with. Let's take a step back.

The following is my understanding of how your PR works. When the prop `accessibilityElementsHidden` is `true`:
  - On iOS: the descendants are hidden
  - On Android: the descendants and the view are hidden

Do you agree with this? If not, can you edit my above description to correct it?

**Comment by omeid** [comment]
Why do you think `importantForAccessiblity=noHideDescendants` would hide the view as well? Please read the code.

**Comment by rigdern** [comment]
@omeid My understanding that it'll hide the view as well came from reading the Android docs for `IMPORTANT_FOR_ACCESSIBILITY_NO_HIDE_DESCENDANTS`: https://developer.android.com/reference/android/view/View.html#IMPORTANT_FOR_ACCESSIBILITY_NO_HIDE_DESCENDANTS

**Comment by omeid** [comment]
You are right, it will hide the view as well, and as I have been using this for High Order Components I have overlooked that difference.

Can you provide an example where this inconsistency would result in inconsistent accessibility behaviour so that me and those interested could look for a solution?

**Comment by sdg9** [comment]
To rigdern's point if `accessibilityElementsHidden={true}` is set on a touchable element, according to what previously was discussed in this PR and RN documentation

> **accessible (iOS, Android)**
> 
> When true, indicates that the view is an accessibility element. When a view is an accessibility element, it groups its children into a single selectable component. By default, all touchable elements are accessible.

while the children and Android touchable element will not be accessible the iOS touchable element will be. (not tested)

What if this PR updates `accessible` to default to false on touchable elements (and anything else defaulting `accessible` to true) when `accessibilityElementsHidden={true}`?

**Comment by omeid** [comment]
@sdg9 I think that is adding too much complexity and making assumptions about developers caring enough about accessibility to set the correct values, setting touchable elements to false will break many multi-cell touchable-(row/list item) components that are fairly common. It will also break many Appium driven tests for said components.

**Comment by omeid** [comment]
I think a better solution is to also set [`isAccessibilityElement`](https://developer.apple.com/reference/objectivec/nsobject/1615141-isaccessibilityelement) to reflect `accessibilityElementsHidden` on iOS to make it react-native's `accessibilityElementsHidden` behave the same on both iOS and Android: Hide the element and it's children.

Thoughts? Hints on implementation reference to follow?

**Comment by sdg9** [comment]
@omeid just to be clear I threw out the idea to default`accessible={false}` on touchable elements **ONLY WHEN** `accessibilityElementsHidden={true}` to behave consistent with android.  If `accessibilityElementsHidden` isn't true toucahble elements would still default `accessible={true}`.  

I agree this would be confusing though.  Basically ADA docs on `accessible` would say

> By default, all touchable elements are accessible... except when they're not.

My use case is like yours I only care about this at the screen level so the difference in behavior between Android and iOS is a non-issue as I'm setting `accessibilityElementsHidden` on a View or ScrollView.

**Comment by omeid** [comment]
@sdg9 Alright, I must have misunderstood your comment then.
So one solution is to sync `isAccessibilityElement ` and `accessibilityElementsHidden` on iOS with react-native's `accessibilityElementsHidden`, which is trivial. 

However, what happens to `accessible`?
Mark `accessible` and ``accessibilityElementsHidden` as conflicting?

**Comment by rigdern** [comment]
> Can you provide an example where this inconsistency would result in inconsistent accessibility behaviour so that me and those interested could look for a solution?

@omeid I don't have a specific real world scenario to share. Given our understanding of the inconsistency of the prop implementations, it shouldn't be too hard to make a contrived one. Whatever solution we end up with here, the prop should have consistent behavior across iOS and Android.

Instead of introducing `accessibilityElementsHidden`, have you considered implementing `importantForAccessibility` for iOS? Based on the discussion here, it sounds like the implementation would look like this:
  - When `importantForAccessibility` is `'yes'`:
    - `isAccessibilityElement = true`
    - `accessibilityElementsHidden = false`
  - When `importantForAccessibility` is `'no'`:
    - `isAccessibilityElement = false`
    - `accessibilityElementsHidden = false`
  - When `importantForAccessibility` is `'no-hide-descendants'`:
    - `isAccessibilityElement = false`
    - `accessibilityElementsHidden = true`
  - When `importantForAccessibility` is `'auto'` or `nulll`:
    - Revert `isAccessibilityElement ` and `accessibilityElementsHidden` to their default values.

This would avoid us having to deprecate `importantForAccessibility`.

**Comment by omeid** [comment]
That is doable, however, the question of what happens to `accessible` prob remains. The `accessible` prob on iOS is effectively `isAccessibilityElement`. Deprecate it? as it doesn't exist on Android anyways.

**Comment by sdg9** [comment]
Heads up @omeid and @rigdern I was playing around with the following

> * When importantForAccessibility is 'yes':
>   * isAccessibilityElement = true
>   * accessibilityElementsHidden = false

and see some unexpected behavior.  Setting isAccessibilityElement to true in iOS will make the View accessible but you cannot drill down into children that should otherwise be accessible (tested on simulator using accessibility inspector).  In android children remain accessible.  

This technically would be a breaking change as people using `yes` may need to switch to `no` or `auto/null` to retain existing functionality in iOS.

**Comment by rigdern** [comment]
> what happens to `accessible` prop

@omeid That's a great question. Given the implementation I mentioned above, the functionality of `accessible` and `importantForAccessibility` overlap on iOS. But on Android, `accessible` and `importantForAccessibility` have no overlap in functionality. This inconsistency in how the props work between iOS and Android is not ideal.

@sdg9 mentioned that the implementation of `importantForAccessibility` that I proposed above results in inconsistent behavior between iOS and Android:

> iOS will make the View accessible but you cannot drill down into children ... In android children remain accessible.

It sounds to me like we need to more deeply understand the iOS and Android accessibility APIs. Then we can use that understanding to implement accessibility props in React Native that offer consistent behvavior across iOS and Android.

**Comment by omeid** [comment]
Consistent behaviour is really important, what is more important is that there _exists_ some/any mechanism that allows one to write reasonably accessible Apps. 

I think over 6 months of bike-shedding on this issue while hundreds to thousands react-native apps that are completely inaccessible are deployed is pretty inconsiderate.

The semantics of accessibility is different on iOS and Android, by the virtue of this difference, it maybe necessary to make some compromises and work the solution out by way of an experimental APIs phase.

Also, on iOS, an accessible element can _not_ have accessible children, so when you set `accessibleElementsHidden=true`, it is assumed that the element is not accessible, which is exactly what happens in Android with `importantForAccessiblity=noHideDescendants`.

So the objection to this pull request being inconsistent is merely a question of difference in semantics, unless, you can provide an example where this inconsistency is manifested in a difference in accessible behaviour.

**Comment by sdg9** [comment]
@omeid I feel your pain

> what is more important is that there exists some/any mechanism that allows one to write reasonably accessible Apps.

At the end of the day all I need to be able to do is set iOS's native property `accessibilityElementsHidden=true` as it will resolve [NavigationExperimental's ADA issue](https://github.com/facebook/react-native/issues/9725) and [react-navigation's ADA issue](https://github.com/react-community/react-navigation/issues/377).  It is logical to align with `importantForAccessibility="no-hide-descendants"` but I think we have to align the iOS and Android behavior around "yes" first.

To summarize the current implementation

| React Native                | iOS                    | Android                   |
|-----------------------------|------------------------|---------------------------|
| accessible                  | isAccessibilityElement | setFocusable              |
| importantForAccessibility   | N/A                    | importantForAccessibility |

and not implemented but discussed

| React Native                | iOS                         | Android |
|-----------------------------|-----------------------------|---------|
| accessibilityElementsHidden | accessibilityElementsHidden | N/A     |


iOS has two booleans to work with: `isAccessibilityElement` and `accessibilityElementsHidden`
Android has `importantForAccessibility` accepting 4 different inputs.

It sounds logical to map the permutations of iOS's 2 booleans to Android's 4 inputs thus avoiding fragmentation in the react native apis.  

Then we discovered 

> on iOS, an accessible element can not have accessible children

which is **NOT** how Android behaves when setting `importantForAccessibility="yes"`.  

I believe we can [alter the suggestion](https://github.com/facebook/react-native/pull/11788#issuecomment-281460644) from @rigdern to achieve consistent behavior between platforms by using iOS's default value for `isAccessibilityElement` when setting `importantForAccessibility="yes"`.

After all [Android's docs](https://developer.android.com/reference/android/view/View.html#IMPORTANT_FOR_ACCESSIBILITY_YES) do not say that `importantForAccessibility="yes"` makes the view accessible but rather

> The view is **important** for accessibility.

This would mean React Native's `importantForAccessibility` prop would never set iOS's `isAccessibilityElement=true`.  If we ever explicitly wanted to set a view as accessible we'd still need to set the `accessible=true` prop.

**Comment by kirankalyan5** [comment]
when can we expect **accessibilityElementsHidden** prop implemented ?
or is there any workaround to hide the accessibility label of previous screens ?

Thanks!

**Comment by lprhodes** [comment]
@omeid It'd be great if you could take a look at this, accessibility should be a real priority

**Comment by sdg9** [comment]
@kirankalyan5 and anyone else still waiting on this as a workaround I found that maintaining my own  React Native fork with this [single commit](https://github.com/sdg9/react-native/commit/a001149ccb6c071a020794861008970038612978) was sufficient for my use case.  Now importantForAccessibility works and behaves as I want in iOS builds.  No more iOS ADA bleed through to previous pages!

I just do something like
```js
importantForAccessibility={this.props.isActive ? 'yes' : 'no-hide-descendants'}
```
to each [card (using react-navigation)](https://github.com/react-community/react-navigation/blob/master/src/views/CardStack/Card.js#L28) so they know when to be accessible or not.

**Comment by lprhodes** [comment]
@sdg9 where's isActive coming from? I don't believe react-navigation supports this yet when pushing cards onto a navigation stack

**Comment by omeid** [comment]
@lprhodes I will be starting a new react native project in a couple of days and will try to chase this again. I agree, this is a massive problem for a18y.

**Comment by sdg9** [comment]
@lprhodes isActive is defined in [react-navigation on the scene](https://github.com/react-community/react-navigation/blob/fe4b1e2379244da8b4909a707760417c06b04452/src/TypeDefinition.js#L409)

I actually had to then pass it to my SceneView [in CardStack](https://github.com/react-community/react-navigation/blob/fe4b1e2379244da8b4909a707760417c06b04452/src/views/CardStack/CardStack.js#L408).  I couldn't use theirs straight up so I made my own tweak changing
```js
      <SceneView
        screenProps={this.props.screenProps}
        navigation={navigation}
        component={SceneComponent}
      />
```
to
```js
      <SceneView
        screenProps={this.props.screenProps}
        navigation={navigation}
        component={SceneComponent}
        isActive={scene.isActive}
      />

```

So in summary in CardStack I had to pass isActive down to the SceneView, and in the Card I had to then use that isActive (on the view) to set importantForAccessibility properly.  Combine that change in react-navigation with [this react-native commit I maintain](https://github.com/sdg9/react-native/commit/a001149ccb6c071a020794861008970038612978), voila I have accessibility working properly without pages on the back stack showing through.

**Comment by lprhodes** [comment]
Thanks @sdg9, that was the missing piece I was looking for. Unfortunately all tabs appear to be active, do you use a work-around for that?

**Comment by sdg9** [comment]
@lprhodes in my use case I don't use TabNavigator, only StackNavigator, so unfortunately I cannot comment on that.

**Comment by mmazzarolo** [comment]
Hi there,
we tried implementing the solution you described above (thanks @sdg9 for your help) but setting the `importantForAccessibility={this.props.scene.isActive ? 'yes' : 'no-hide-descendants'}` was not enough for not leaking the touch interactions: when more then one scene was rendered, the touch interaction would be captured by both the active and inactive scenes.  

Since our app is highly focused on handling the accessibility for unpaired users, we opted for changing the `_renderCard` function in the `react-navigation/src/views/CardStack/CardStack.js`:  

```javascript
_renderCard = (scene: NavigationScene): React.Node => {
    const { screenInterpolator } = this._getTransitionConfig();
    const style = screenInterpolator && screenInterpolator({ ...this.props, scene });

    const SceneComponent = this.props.router.getComponentForRouteName(scene.route.routeName); 
    // We do not render the views down the stack except the active one.
    // Be aware that this is an horrible solution for fixing the leaking touch interactions
    // that break the accessibility.
    if (!scene.isActive) {
      return <View key={`card_${scene.key}`} />;
    }
    return (
      <Card
        {...this.props}
        key={`card_${scene.key}`}
        style={[style, this.props.cardStyle]}
        scene={scene}
        importantForAccessibility={scene.isActive ? 'yes' : 'no-hide-descendants'}
        accessible={false}
      >
        {this._renderInnerScene(SceneComponent, scene)}
      </Card>
    );
  };
```
Hope it helps.

cc @andreatosatto90

**Comment by iacopo87** [comment]
+1

**Comment by awseeley** [comment]
Hi - I implemented the fix from @mmazzarolo. However as I was using animations between stack navigators this broke my animations. 

I have applied the following fix that will work with animations & also fix talkback from reading components on inactive screens in the stack navigator. To fix it I have changed the render method in the following file:

`react-navigation/src/views/CardStack/CardStack.js`

`  render() {
    const { children, pointerEvents, style } = this.props;

    if(this.props.scene.isActive) {
      importantForAccessibility='yes';
    } else {
      importantForAccessibility='no-hide-descendants';
    }
    
    return (
      <Animated.View
        pointerEvents={pointerEvents}
        ref={this.props.onComponentRef}
        style={[styles.main, style]}
        importantForAccessibility={importantForAccessibility}
      >
        {children}
      </Animated.View>
    );
  }
}`

This fix has also been forked on github & published to NPM if anyone needs it as follows:
[https://github.com/awseeley/react-navigation](url)
[https://www.npmjs.com/package/react-navigation-awseeley](url)

**Comment by dryganets** [comment]
This PR could be closed as functionality you are trying to add is already implemented in master:
https://github.com/facebook/react-native/commit/31288161e188723456fdb336c494f3c8a3f5b0a8

**Comment by omeid** [comment]
Unfort because of lack traction with landing this issue and me having moved on from react-native for a while now. I am going to close it.
